File: odoo_surface_mcp/cache.py
"""Simple in-memory cache with optional TTL and hit/miss tracking."""
import time
from typing import Any


class Cache:
    def __init__(self, default_ttl: int = 300):
        """
        Args:
            default_ttl: seconds before an entry expires (0 = never)
        """
        self._store: dict[str, tuple[Any, float]] = {}  # key → (value, expires_at)
        self._hits = 0
        self._misses = 0
        self.default_ttl = default_ttl

    # ------------------------------------------------------------------
    # Core operations
    # ------------------------------------------------------------------

    def get(self, key: str) -> Any:
        """Return cached value or None if missing/expired."""
        entry = self._store.get(key)
        if entry is None:
            self._misses += 1
            return None
        value, expires_at = entry
        if expires_at and time.time() > expires_at:
            del self._store[key]
            self._misses += 1
            return None
        self._hits += 1
        return value

    def set(self, key: str, value: Any, ttl: int | None = None) -> None:
        """Store value under key. ttl=0 or ttl=None uses default_ttl."""
        effective_ttl = ttl if ttl is not None else self.default_ttl
        expires_at = (time.time() + effective_ttl) if effective_ttl else 0.0
        self._store[key] = (value, expires_at)

    def clear(self) -> int:
        """Remove all entries. Returns count cleared."""
        count = len(self._store)
        self._store.clear()
        self._hits = 0
        self._misses = 0
        return count

    # ------------------------------------------------------------------
    # Diagnostics
    # ------------------------------------------------------------------

    def stats(self) -> dict:
        now = time.time()
        live = sum(1 for _, (_, exp) in self._store.items() if not exp or exp > now)
        expired = len(self._store) - live
        return {
            "entries_live": live,
            "entries_expired": expired,
            "hits": self._hits,
            "misses": self._misses,
            "hit_rate": round(
                self._hits / (self._hits + self._misses), 3
            ) if (self._hits + self._misses) else 0.0,
        }

    def dump(self) -> list[dict]:
        """Return all live entries with key and expiry info (no values)."""
        now = time.time()
        rows = []
        for key, (_, exp) in self._store.items():
            if exp and exp <= now:
                continue
            rows.append({
                "key": key,
                "expires_in_s": round(exp - now, 1) if exp else None,
            })
        return sorted(rows, key=lambda r: r["key"])

File: src/cache.ts
/** Simple in-memory cache with TTL and hit/miss tracking. */

export interface CacheStats {
  entries_live: number;
  entries_expired: number;
  hits: number;
  misses: number;
  hit_rate: number;
}

export interface CacheDumpEntry {
  key: string;
  expires_in_s: number | null;
}

interface Entry {
  value: unknown;
  expiresAt: number; // ms epoch; 0 = never
}

export class Cache {
  private store = new Map<string, Entry>();
  private hits = 0;
  private misses = 0;

  constructor(public readonly defaultTtl: number = 300) {}

  get(key: string): unknown {
    const entry = this.store.get(key);
    if (!entry) { this.misses++; return undefined; }
    if (entry.expiresAt && Date.now() > entry.expiresAt) {
      this.store.delete(key);
      this.misses++;
      return undefined;
    }
    this.hits++;
    return entry.value;
  }

  set(key: string, value: unknown, ttl?: number): void {
    const effective = ttl ?? this.defaultTtl;
    const expiresAt = effective ? Date.now() + effective * 1000 : 0;
    this.store.set(key, { value, expiresAt });
  }

  clear(): number {
    const count = this.store.size;
    this.store.clear();
    this.hits = 0;
    this.misses = 0;
    return count;
  }

  stats(): CacheStats {
    const now = Date.now();
    let live = 0;
    let expired = 0;
    for (const { expiresAt } of this.store.values()) {
      if (!expiresAt || expiresAt > now) live++;
      else expired++;
    }
    const total = this.hits + this.misses;
    return {
      entries_live: live,
      entries_expired: expired,
      hits: this.hits,
      misses: this.misses,
      hit_rate: total ? Math.round((this.hits / total) * 1000) / 1000 : 0,
    };
  }

  dump(): CacheDumpEntry[] {
    const now = Date.now();
    const rows: CacheDumpEntry[] = [];
    for (const [key, { expiresAt }] of this.store.entries()) {
      if (expiresAt && expiresAt <= now) continue;
      rows.push({
        key,
        expires_in_s: expiresAt ? Math.round((expiresAt - now) / 100) / 10 : null,
      });
    }
    return rows.sort((a, b) => a.key.localeCompare(b.key));
  }
}
